openmc/surface.py

```diff
@@ -268,62 +293,103 @@ def from_hdf5(group):
-            elif surf_type == 'y-cone':
-                surface = YCone(x0, y0, z0, r2, bc, name, surface_id)
-            elif surf_type == 'z-cone':
-                surface = ZCone(x0, y0, z0, r2, bc, name, surface_id)
-
-        elif surf_type == 'quadric':
-            a, b, c, d, e, f, g, h, j, k = coeffs
-            surface = Quadric(a, b, c, d, e, f, g, h, j, k, bc, name, surface_id)
+        cls = _SURFACE_CLASSES[surf_type]
+
+        return cls(*coeffs, **kwargs)
+
+
+class PlaneMixin(metaclass=ABCMeta):
+    """A Plane mixin class for all operations on order 1 surfaces"""
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._periodic_surface = None
+
+    @property
+    def periodic_surface(self):
+        return self._periodic_surface
+
+    @periodic_surface.setter
+    def periodic_surface(self, periodic_surface):
+        check_type('periodic surface', periodic_surface, Plane)
+        self._periodic_surface = periodic_surface
+        periodic_surface._periodic_surface = self
+
+    @abstractmethod
+    def _get_base_coeffs(self):
+        """Return coefficients a, b, c, d representing a general plane of the
+        form :math:`ax + by + cz = d`.
+
+        """
+        pass
+
+    def evaluate(self, point):
+        """Evaluate the surface equation at a given point.
+
+        Parameters
+        ----------
+        point : 3-tuple of float
+            The Cartesian coordinates, :math:`(x',y',z')`, at which the surface
+            equation should be evaluated.
+
+        Returns
+        -------
+        float
+            :math:`Ax' + By' + Cz' - D`
+
+        """
+
+        x, y, z = point
+        a, b, c, d = self._get_base_coeffs()
+        return a*x + b*y + c*z - d
+
+    @abstractmethod
+    def translate(self, vector, clone=True):
+        """Translate surface in given direction
+
+        Parameters
+        ----------
+        vector : iterable of float
+            Direction in which surface should be translated
+        clone : boolean
+            Whether or not to return a new instance of a Plane or to modify the
+            coefficients of this plane. Defaults to True
 
-        return surface
+        Returns
+        -------
+        openmc.Plane
+            Translated surface
+
+        """
+        pass
+
+    def to_xml_element(self):
+        """Return XML representation of the surface
+
+        Returns
+        -------
+        element : xml.etree.ElementTree.Element
+            XML element containing source data
+
+        """
+        element = super().to_xml_element()
+
+        # Add periodic surface pair information
+        if self.boundary_type == 'periodic':
+            if self.periodic_surface is not None:
+                element.set("periodic_surface_id",
+                            str(self.periodic_surface.id))
+        return element
 
 
-class Plane(Surface):
+class Plane(PlaneMixin, Surface):
```

Comment:
Since this mixin has the `to_xml_element` code for Planes it needs to come first in the MRO for cooperative inheritance to work, which seems to make sense to me as it is intermediate in the hierarchy between a concrete Plane object and a general surface. It's how I prevent XPlanes etc  from inheriting the attributes of Plane which caused inconsistencies in the first place. The alternative would be to get rid of the mixin, inherit from Plane and override setters to disallow setting properties like a, b,c, d but the mixin seemed like the more elegant solution, but I'd be interested in hearing your thoughts.